### PR TITLE
[batchnorm] Optimize batch norm for memory and speed @open sesame 09/28 18:42

### LINKED ISSUE #1598: Q. Shouldn't CalcGradient be skipped when non-trainable?
https://github.com/nnstreamer/nntrainer/blob/51a871bd7b6e63ddcac6185e55ae802c804d1d70/nntrainer/models/neuralnet.cpp#L244-L245

Current upstream code does not skip calcGradient when `node->trainable() == false`, I wonder if this is because some calculation is depending on calcGradient() (which they shouldn't)?

### PR DESCRIPTION
**Pull request by kparichay**
- This patch optimizes batch normalization implementation.
Reduces the temporary memory allocation and provide speedup for the
layer execution. With this patch, resnet18 runtime improves by approx
5%.
  - update the tensor usage method to give the output tensor. Although the methods where already allocated output can be passed were available, they were not being used. This patch adds the corresponding support.
- This patch optimizes batch norm layer and tries to share the
calculations performed in calcGradient and calcDerivative.
    - reuse dbeta and dgamma calculations
    - reduce number of required temporary variables
    - create all the required tensor variables with context
    - add support for checking if the layer is trainable or not via run
    context
    - support average operation with the output tensor already allocated
    - this patch reduces as much as memory as possible without sacrificing
    speed. more memory optimization is possible at the expense of speed but
    has been ommitted for now.

See also  #1591

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

resolves #1598

### CODE REVIEW

**Comment by kparichay** [comment]
Rebase with #1584 and ensure that batch size 2 unittest is passing, and also check for batch size 4.

File: nntrainer/layers/bn_layer.cpp [hunk 1]
```diff
@@ -89,6 +110,18 @@ void BatchNormalizationLayer::finalize(InitLayerContext &context) {
...
+  wt_idx[BNParams::invstd] =
+    context.requestTensor(dim, context.getName() + ":invstd",
+                          Tensor::Initializer::NONE, false, ITERATION_LIFESPAN);
+  wt_idx[BNParams::t_reduced] =
+    context.requestTensor(dim, context.getName() + ":tensor_reduced",
+                          Tensor::Initializer::NONE, false, ITERATION_LIFESPAN);
+  wt_idx[BNParams::t_full_fw] = context.requestTensor(
+    in_dim, context.getName() + ":tensor_full_fw", Tensor::Initializer::NONE,
+    false, FORWARD_FUNC_LIFESPAN);
+  wt_idx[BNParams::t_full_bw] = context.requestTensor(
```
Comment (jijoongmoon) [inline nntrainer/layers/bn_layer.cpp:122]: Just for checking, if we set FORWARD_FUNC_LIFESPAN here, is it null in backwarding?

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/layers/bn_layer.cpp [hunk 2]
```diff
@@ -147,35 +185,49 @@ void BatchNormalizationLayer::calcDerivative(RunLayerContext &context) {
...
-  Tensor dx_1 = gamma.multiply(invstd);
-  Tensor dx_2 = deriv.multiply(N);
-  dx_2.subtract_i(deriv.sum(axes_to_reduce));
-  dx_2.subtract_i(deviation.divide(cvar).multiply(
-    deviation.multiply(deriv).sum(axes_to_reduce)));
+  dx = deriv.subtract(t_reduced);
 
-  dx = dx_2.multiply(dx_1, dx);
-  dx.divide_i(N);
+  t_full = deviation.multiply(deriv);
```
Comment (zhoonit) [inline nntrainer/layers/bn_layer.cpp:206]: deviation.multiply(deriv, t_full); ?

File: nntrainer/layers/bn_layer.cpp [hunk 3]
```diff
@@ -147,35 +185,49 @@ void BatchNormalizationLayer::calcDerivative(RunLayerContext &context) {
...
-  Tensor &dgamma = context.getWeightGrad(wt_idx[BNParams::gamma]);
   Tensor &dbeta = context.getWeightGrad(wt_idx[BNParams::beta]);
   Tensor &deriv = context.getIncomingDerivative(SINGLE_INOUT_IDX);
-  Tensor &deviation = context.getTensor(wt_idx[BNParams::deviation]);
 
-  dbeta = deriv.sum(axes_to_reduce);
-  Tensor dev = deviation.multiply(invstd);
-  dev.multiply_i(deriv);
-  dgamma = dev.sum(axes_to_reduce);
+  deriv.sum(axes_to_reduce, dbeta);
```
Comment (zhoonit) [inline nntrainer/layers/bn_layer.cpp:230]: only beta is calculated here and gamma is calculated in calcDeriv?

File: nntrainer/layers/bn_layer.cpp [hunk 4]
```diff
@@ -147,35 +185,49 @@ void BatchNormalizationLayer::calcDerivative(RunLayerContext &context) {
...
+  } else {
+    t_reduced = deriv.average(axes_to_reduce);
   }
 
-  Tensor dx_1 = gamma.multiply(invstd);
-  Tensor dx_2 = deriv.multiply(N);
-  dx_2.subtract_i(deriv.sum(axes_to_reduce));
-  dx_2.subtract_i(deviation.divide(cvar).multiply(
-    deviation.multiply(deriv).sum(axes_to_reduce)));
+  dx = deriv.subtract(t_reduced);
```
Comment (zhoonit) [inline nntrainer/layers/bn_layer.cpp:204]: deriv.subtract(t_reduced, dx)?

File: nntrainer/layers/bn_layer.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/layers/bn_layer.cpp:122]: currently, it is in available backward (run context doesn't know when function is calling it), but later we will have a check so `context.requestTensor()` will throw in this case.

Comment (kparichay) [reply nntrainer/layers/bn_layer.cpp:122]: I have added a corresponding todo in layer context.

File: nntrainer/layers/bn_layer.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/layers/bn_layer.cpp:206]: good catch, done.

File: nntrainer/layers/bn_layer.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/layers/bn_layer.cpp:204]: good catch, done.

File: nntrainer/layers/bn_layer.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/layers/bn_layer.cpp:230]: yes.

File: nntrainer/layers/bn_layer.cpp [hunk 5]
```diff
@@ -89,6 +110,18 @@ void BatchNormalizationLayer::finalize(InitLayerContext &context) {
   wt_idx[BNParams::deviation] =
     context.requestTensor(in_dim, context.getName() + ":deviation",
                           Tensor::Initializer::NONE, false, ITERATION_LIFESPAN);
+  wt_idx[BNParams::invstd] =
+    context.requestTensor(dim, context.getName() + ":invstd",
+                          Tensor::Initializer::NONE, false, ITERATION_LIFESPAN);
+  wt_idx[BNParams::t_reduced] =
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/layers/bn_layer.cpp:116]: added in the comments.

**Review by zhoonit — APPROVED** [review]
LGTM!
